### PR TITLE
Parallelizes ANK snippet compilation on a thread per file.

### PR DESCRIPTION
**Pull request by raulraja**

### CODE REVIEW

File: ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt [hunk 1]
```diff
@@ -118,37 +126,58 @@ fun extractCodeImpl(source: String, tree: ASTNode): ListKW<Snippet> {
...
+        }
+    }
+    return runBlocking {
+        snippets.map { (file, codeBlocks) ->
+            async(IOPool) {
+                val start = System.currentTimeMillis()
+                val classLoader = URLClassLoader(classpath.map { URL(it) }.ev().list.toTypedArray())
+                val seManager = ScriptEngineManager(classLoader)
+                println(":runAnk -> START eval: $file on ${Thread.currentThread().name}")
+                val engineCache: Map<String, ScriptEngine> =
```
Comment (pakoito) [inline ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt]: Should the engine cache be local per script, or shared among all of them?

File: ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt [hunk 2]
```diff
@@ -118,37 +126,58 @@ fun extractCodeImpl(source: String, tree: ASTNode): ListKW<Snippet> {
...
-    val evaledSnippets = snippets.list.map { snippet ->
-        val result = Try {
-            val engine = cachedEngines.get(snippet.lang)
-            val resolvedEngine = if (engine != null) engine else {
-                seManager.getEngineByExtension(extensionMappings.getOrDefault(snippet.lang, "kts"))!!
+
+fun compileCodeImpl(snippets: Map<File, ListKW<Snippet>>, classpath: ListKW<String>): ListKW<CompiledMarkdown> {
+    println(":runAnk -> started compilation")
+    val IOPool = object : CoroutineDispatcher() {
+        private val ioBoundPool = Executors.newFixedThreadPool(snippets.size)
```
Comment (pakoito) [inline ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt]: One Thread per file?

File: ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt [hunk 3]
```diff
@@ -118,37 +126,58 @@ fun extractCodeImpl(source: String, tree: ASTNode): ListKW<Snippet> {
...
-    val cachedEngines: ConcurrentHashMap<String, ScriptEngine> = ConcurrentHashMap()
-    val evaledSnippets = snippets.list.map { snippet ->
-        val result = Try {
-            val engine = cachedEngines.get(snippet.lang)
-            val resolvedEngine = if (engine != null) engine else {
-                seManager.getEngineByExtension(extensionMappings.getOrDefault(snippet.lang, "kts"))!!
+
+fun compileCodeImpl(snippets: Map<File, ListKW<Snippet>>, classpath: ListKW<String>): ListKW<CompiledMarkdown> {
+    println(":runAnk -> started compilation")
+    val IOPool = object : CoroutineDispatcher() {
```
Comment (pakoito) [inline ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt]: I'm quite sure that kotlinx.coroutines provides a better abstraction for a weighted ThreadPool.

**Comment by nomisRev** [comment]
Gets stuck on output. I get same result locally. Seems to work fine on `CommonPool`, I also tried `newSingleThreadExecutor` I wanted to monitor difference with `CommonPool` but on `newSingleThreadExecutor` it also hang on output.

File: ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt (hunk 1 above)
Comment (raulraja) [reply ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt]: I was told they were thread safe but those rx issues I mentioned to you in private show they are not so there needs to be one instance per file basically. Those engines are not thread safe.

**Comment by raulraja** [comment]
@nomisRev it eventually completes, this concurrent execution does not hide the fact that the Kotlin Script Engine is broken when evaling corotuines and other code that does type fiddling more than it is used to. If you connect with a debugger and pause the execution you will find yourself inside a `while(true)` in the Script Engine impl that is trying to resolve variable scopes and never ending until it gives up eventually.

**Comment by nomisRev** [comment]
@raulraja even while it is printing the output files?

**Pull request merged**